numba/targets/imputils.py

```diff
@@ -72,6 +72,23 @@ def imp(context, builder, sig, args):
             builder, func, fndesc.restype, fndesc.argtypes, args, env=None)
         with cgutils.if_unlikely(builder, status.is_error):
             context.call_conv.return_status_propagate(builder, status)
+        assert sig.return_type == fndesc.restype
+        # Reconstruct optional return type
+        if isinstance(sig.return_type, types.Optional):
```

Comment:
numba-compiled function unpacks optional return type and uses return-code to indicate whether it returns None or the value.  As a result the LLVM value of the returned value is actually not a optional type.  It is legacy code before the introduction of the ``Optional``.  Perhaps, we should unify everything to use Optional and remove the return-code for None.